### ISSUE TITLE
alter index html response

### ISSUE THREAD

**Issue by samselikoff** [tedconf/front_end_builds#33]
Just jotting an idea down, Luke mentioned in his talk they sometimes alter the HTML response after it was fetched from Redis. For example, they inject `current_user` into a meta tag, so the Ember app doesn't have to make an XHR request.

We may want to eventually have an API for this type of thing.

**Comment by hhff** [comment]
was just about to open an issue for this - this would be amazing

**Comment by hhff** [comment]
Perhaps the developer simply defines a hash in the app controller to be converted to meta tags

```ruby
def feb_meta(app_name)
    if app_name == "blog"
        { current_user: current_user }
    end
end
```

**Comment by samselikoff** [comment]
Just brainstorming with Ryan about this, one way we could design this would be to have user-defined transforms.

```rb
class MyTransformer < FrontEndBuilds::Transformer
  def transforms(scope, html)
    append_to_head(html, "<meta value=#{scope.current_user.to_json}")
  end
end

FrontEndBuilds.register_transforms(MyTransformer)

# every request, instantiate the transformers, and call #transform on the scope and html
```

`add_to_head` could be a helper method provided by FEB, which adds the string to the end of the `<head>` tag in the index html (which was fetched from the db). You could similarly define `prepend_to_head`, `append_to_body` and `prepend_to_body` methods.

This way, FEB could provide a CSRF transformer that works out the box (you could disable it), and we could apply that transform in addition to any user-generated ones.

**Comment by ngottlieb** [comment]
Hmm, I like that idea, but I wonder if it isn't more complex than is needed. @hhff's solution covers all of my needs and strikes me as a lot simpler and I would guess covers the vast majority of people's needs.

**Comment by samselikoff** [comment]
Yep, great point. I think this covers all known use cases for now, and much simpler!

**Comment by ngottlieb** [comment]
Cool -- I'll try to get this sorted and submit a PR this week sometime.

**Comment by hhff** [comment]
i tend to think it would be worth exposing a low level hook that the index.html string gets passed through for the developer to override and do whatever manipulation they want too.  maybe that's where we call the `setup_metatags` hook?

**Comment by ngottlieb** [comment]
I wonder if it isn't worth making that a separate feature, though? Just because the mechanics for meta tags is already there, and it's a very simple and common requirement.

**Comment by hhff** [comment]
yeah, i mean it's just 3 lines of code to add the low level hook also

**Comment by ngottlieb** [comment]
I guess I could go either way -- I'm relatively new to this gem, so I'll defer to you guys.

**Comment by samselikoff** [comment]
Why don't we start with a `feb_meta` method that returns a hash

**Comment by ngottlieb** [comment]
How are you guys envisioning that a user/developer overrides `feb_meta`? I'd love to do something like Devise does so you could call super and incorporate the default meta tags + your own, but the way Devise makes that work involves a lot of overhead that I'm sure we don't want. So I just tried implementing a solution to that using Rails' decorator motif and it doesn't seem great. It doesn't end up saving the developer much time over just decorating the `BestsController` method and overriding `meta_tags`.

https://github.com/ngottlieb/front_end_builds/tree/feb_meta

Thoughts? It could be pretty neat to go further with this and expose setting custom meta variables to the UI per app or per build.

**Comment by samselikoff** [comment]
I think what you have solves your need, and we're happy to merge it in. We can revisit it later as more needs arise - I guess we don't have a ton of time just this moment to look further into this. How does that sound?

**Comment by ngottlieb** [comment]
Works for me, I'll submit a PR. I'd be happy to do more work on it with a bit of direction -- whether that's towards exposing customizable meta tags to the UI or building the Transformers idea out -- I think both are useful ideas.
